nomad/operator_endpoint.go

```diff
@@ -470,3 +473,155 @@ func (op *Operator) snapshotSave(conn io.ReadWriteCloser) {
+		handleFailure(500, err)
+		return
+	}
+
+	// Forward to appropriate region
+	if args.Region != op.srv.Region() {
+		err := op.forwardStreamingRPC(args.Region, "Operator.SnapshotRestore", args, conn)
+		if err != nil {
+			handleFailure(500, err)
+		}
+		return
+	}
+
+	// forward to leader
+	remoteServer, err := op.srv.getLeaderForRPC()
+	if err != nil {
+		handleFailure(500, err)
+		return
+	}
+	if remoteServer != nil {
+		err := op.forwardStreamingRPCToServer(remoteServer, "Operator.SnapshotRestore", args, conn)
+		if err != nil {
+			handleFailure(500, err)
+		}
+		return
+
+	}
+
+	// Check agent permissions
+	if aclObj, err := op.srv.ResolveToken(args.AuthToken); err != nil {
+		code := 500
+		if err == structs.ErrTokenNotFound {
+			code = 400
+		}
+		handleFailure(code, err)
+		return
+	} else if aclObj != nil && !aclObj.IsManagement() {
+		handleFailure(403, structs.ErrPermissionDenied)
+		return
+	}
+
+	op.srv.setQueryMeta(&reply.QueryMeta)
+
+	reader, errCh := decodeStreamOutput(decoder)
+
+	err = snapshot.Restore(op.logger.Named("snapshot"), reader, op.srv.raft)
+	if err != nil {
+		handleFailure(500, fmt.Errorf("failed to restore from snapshot: %v", err))
+		return
+	}
+
+	err = <-errCh
+	if err != nil {
+		handleFailure(400, fmt.Errorf("failed to read stream: %v", err))
+		return
+	}
+
+	// This'll be used for feedback from the leader loop.
+	timeoutCh := time.After(time.Minute)
+
+	lerrCh := make(chan error, 1)
+
+	select {
+	// Reassert leader actions and update all leader related state
+	// with new state store content.
+	case op.srv.reassertLeaderCh <- lerrCh:
+
+	// We might have lost leadership while waiting to kick the loop.
+	case <-timeoutCh:
+		handleFailure(500, fmt.Errorf("timed out waiting to re-run leader actions"))
+
+	// Make sure we don't get stuck during shutdown
+	case <-op.srv.shutdownCh:
+	}
+
+	select {
+	// Wait for the leader loop to finish up.
+	case err := <-lerrCh:
+		if err != nil {
+			handleFailure(500, err)
+			return
+		}
+
+	// We might have lost leadership while the loop was doing its
+	// thing.
+	case <-timeoutCh:
+		handleFailure(500, fmt.Errorf("timed out waiting for re-run of leader actions"))
+
+	// Make sure we don't get stuck during shutdown
+	case <-op.srv.shutdownCh:
+	}
+
+	reply.Index, _ = op.srv.State().LatestIndex()
+	op.srv.setQueryMeta(&reply.QueryMeta)
+	encoder.Encode(reply)
+}
+
+func decodeStreamOutput(decoder *codec.Decoder) (io.Reader, <-chan error) {
+	pr, pw := io.Pipe()
+	errCh := make(chan error, 1)
+
+	go func() {
+		defer close(errCh)
+
+		for {
+			var wrapper cstructs.StreamErrWrapper
+
+			err := decoder.Decode(&wrapper)
+			if err != nil {
+				pw.CloseWithError(fmt.Errorf("failed to decode input: %v", err))
+				errCh <- err
+				return
+			}
+
+			if len(wrapper.Payload) != 0 {
+				_, err = pw.Write(wrapper.Payload)
+				if err != nil {
+					pw.CloseWithError(err)
+					errCh <- err
+				}
+			}
+
+			if errW := wrapper.Error; errW != nil {
```

Comment:
Can `wrapper` have both a Payload and an Error? I suppose that makes sense for signalling EOF while also sending the final payload.

Reply:
It doesn't.  A wrapper is currently either a Payload or an error.  I ensured that we handle EOF by sending payload separately.  However, I used the "be flexible in what you take" approach in the rpc client side just in case of future changes, and to ease anxiety of future readers of the code (who may ask the other question and think there is a bug).